extmod/vfs_fat_misc.c

```diff
@@ -34,51 +34,64 @@
 #include "extmod/vfs_fat.h"
 #include "py/lexer.h"
 
-// TODO: actually, the core function should be ilistdir()
-
-mp_obj_t fat_vfs_listdir2(fs_user_mount_t *vfs, const char *path, bool is_str_type) {
-    FRESULT res;
-    FILINFO fno;
+typedef struct _mp_vfs_fat_ilistdir_it_t {
+    mp_obj_base_t base;
+    mp_fun_1_t iternext;
+    bool is_str;
     FF_DIR dir;
+} mp_vfs_fat_ilistdir_it_t;
 
-    res = f_opendir(&vfs->fatfs, &dir, path);
-    if (res != FR_OK) {
-        mp_raise_OSError(fresult_to_errno_table[res]);
-    }
-
-    mp_obj_t dir_list = mp_obj_new_list(0, NULL);
+STATIC mp_obj_t mp_vfs_fat_ilistdir_it_iternext(mp_obj_t self_in) {
+    mp_vfs_fat_ilistdir_it_t *self = MP_OBJ_TO_PTR(self_in);
 
     for (;;) {
-        res = f_readdir(&dir, &fno);                   /* Read a directory item */
-        if (res != FR_OK || fno.fname[0] == 0) break;  /* Break on error or end of dir */
-        if (fno.fname[0] == '.' && fno.fname[1] == 0) continue;             /* Ignore . entry */
-        if (fno.fname[0] == '.' && fno.fname[1] == '.' && fno.fname[2] == 0) continue;             /* Ignore .. entry */
-
+        FILINFO fno;
+        FRESULT res = f_readdir(&self->dir, &fno);
         char *fn = fno.fname;
+        if (res != FR_OK || fn[0] == 0) {
+            // stop on error or end of dir
+            break;
+        }
+        if (fn[0] == '.' && (fn[1] == 0 || (fn[1] == '.' && fn[2] == 0))) {
```

Comment:
I was going by CPython's os.listdir(), which *doesn't* return . and ...  Neither does os.scandir().  And because . and .. are specific to the VFS (probably VfsSPIFFS , if it existed, wouldn't have such entries).  So I think it's up to the VFS to discard them.